### ISSUE TITLE
An explicit type for a generic intrinsic function should not be allowed

### ISSUE THREAD

**Issue by klausler** [j3-fortran/fortran_proposals#198]
This program conforms to Fortran 2018:

```
integer :: cos
print *, cos(3.14159)
end
```

and prints `-1.00000`, or some other amount of zeroes, with every Fortran compiler to which I have access.  The call to `cos()` returns a REAL no matter what the explicit type declaration may be.  One compiler emits a warning.

An explicit type-declaration-stmt for a generic intrinsic function is defined to have "no effect" (subclause 8.2, paragraph 3).

This behavior is unintuitive at best, useful only for stumping Fortran programmers as trivia question, and may even perhaps dangerous -- implementations are free to extend their lists of intrinsic procedures, and a name clash could silently break existing code that uses an implicit interface.

This feature should be deleted.  Existing programs with currently ignored type declarations could be trivially brought up to date without ambiguity or loss of compatibility by simply deleting the declarations.

**Comment by klausler** [comment]
The standards from FORTRAN '77 through Fortran 2008 all had language in them about how a type-declaration-stmt for a generic intrinsic function would not remove the generic nature of the name; the language about the type-declaration-stmt having "no effect" on a generic intrinsic is a recent update in Fortran 2018.

**Comment by jeffhammond** [comment]
Isn't this a specific case of the general issue with Fortran not having any reserved identifiers, which allows programmers to do awful things like define a variable with the same name as an intrinsic function?

Or are you specifically wanting to limit to the change of type from `integer` to `real`, or similar?  What about `integer*4` and `integer*8`?  Or declaring a variable `len` that is an array of integers?

I found numerous uses of variable names that are the same as intrinsics in quantum chemistry codes, although none so far changing the type from `real` to `integer` or vice versa.

**Comment by klausler** [comment]
No, it's not.  The standard specifically says that type declarations on intrinsic function names are ignored, even when they're still being used as intrinsic functions.  So

```
character acos
intrinsic acos
print *, acos(1.)
```

is a conforming program because the type declaration is completely ignored and `ACOS(1.)` is still real.

**Comment by klausler** [comment]
> I am unsure what the original post is trying to achieve and what is the proposal.

The original post was quite clear: *This feature should be deleted.*  The noun phrase "this feature" refers to allowing and ignoring type declarations of names that are generic intrinsic functions.  "Deleted" is a past participle of a verb meaning to remove. expunge, excise, omit, or obliterate.  Put the two together: the standard should no longer ignore a type declaration of a generic intrinsic function.

**Comment by FortranFan** [comment]
To me, the original post is unclear.

Also, my understanding is the posted snippet in the comment upthread does not conform as a program:
https://github.com/j3-fortran/fortran_proposals/issues/198#issuecomment-1170105475

Consider a variant of the snippet:
```fortran
    character acos
   intrinsic acos
end
```
I believe section 19 of the standard makes this variant nonconforming.  That processors are not required to diagnose and detect the nonconformance and few, if any, report any nonconformance is another matter,

But again, I fail to understand the proposal here - there is nothing to delete here.  One can request the J3 committee on this, but I won't be surprised at all if they too take no action on this.

**Comment by klausler** [comment]
In F'2018, subclause 8.2, paragraph 3, you can read "Specifying a type for a generic intrinsic function name in a type declaration statement has no effect."  I invite you to write your own conforming program that specifies a type for a generic intrinsic function name in a type declaration.

**Comment by FortranFan** [comment]
> In F'2018, subclause 8.2, paragraph 3, you can read "Specifying a type for a generic intrinsic function name in a type declaration statement has no effect." I invite you to write your own conforming program that specifies a type for a generic intrinsic function name in a type declaration.

Again, what is the point?  If the proposal is better wording in the second sentence in that subclause or in that paragraph or section, one can submit a "paper" (as defined by J3) and see how it moves forward.  Good or bad, it is an extremely high bar generally with the committee and the editor specifically to proceed with wording changes, but one can try.

On the other hand, if the proposal is something to be "deleted," it is entirely unclear what specifically and what edits are to be made to the standard as a result.  Then there is the issue of backward compatibility that the committee will want to consider by themselves and place a high bar to overcome.

Yet again if the proposal to add a numbered constraint that seeks to prohibit a type declaration with a generic interface name and an intrinsic statement toward such a generic intrinsic procedure in a given scope, one can submit a "paper" toward this and see how it gets received.  Chances are high the committee will deem it unnecessary.
